---
title: "R Notebook"
output: 
---

```{r}
library(data.table)
library(tidyverse)
theme_set(theme_classic())
mane_gtf = rtracklayer::import('annotations/MANE.GRCh38.v1.0.ensembl_genomic.gff.gz')
mane_dt = as.data.table(mane_gtf)

mane_ann = fread('annotations/MANEv1.0.txt'  )
spliceai_ann = fread('annotations/grch38.txt')
```


We found that the annotation file used for the precomputed scores included only a short isoform of DNMT3A which excludes this variant. This short isoform excludes 23.3% of the CDS of the MANE isoform, as well as y% of pathogenic ClinVar variants reported in this gene.

```{r}
precomp_dnmt3a_start = spliceai_ann[`#NAME` == 'DNMT3A', TX_START]
precomp_dnmt3a_end = spliceai_ann[`#NAME` == 'DNMT3A', TX_END]

mane_dt[type == 'CDS' & gene_name == 'DNMT3A' & end >  precomp_dnmt3a_end]
sum(mane_dt[type == 'CDS' & gene_name == 'DNMT3A' & end >  precomp_dnmt3a_end, width]) / 
  sum(mane_dt[type == 'CDS' & gene_name == 'DNMT3A', width])


gnomad_dnmt3a = fread('gnomad_data/variants.csv', 
                      col.names = c('clinvar_id', 'variant_id', 
                                    'gold_start', 'clin_significance', 'review_status'))

gnomad_dnmt3a[, var_pos := as.numeric(sapply(strsplit(variant_id, split = '-', fixed = TRUE), '[[', 2))]

nrow(gnomad_dnmt3a[var_pos > precomp_dnmt3a_end  & clin_significance  %in% c("Likely pathogenic", "Pathogenic/Likely pathogenic", "Pathogenic")]) / nrow(gnomad_dnmt3a[ clin_significance  %in% c("Likely pathogenic", "Pathogenic/Likely pathogenic", "Pathogenic")])

```

# overall- get count of missing gens, missing exons (CDS/UTR?), changed CDS exon boundaries, changed UTR boundaries


```{r}
# annotate with gene names
mane_ann = mane_dt[type == 'transcript', .(`#NAME` = transcript_id, gene_name)][mane_ann, on = .(`#NAME`)]


#making table to find missing segemnts in spliceai annotations
mane_ann_exons = mane_ann[, .(mane_exon_start = as.numeric(unlist(strsplit(EXON_START, split = ','))),
                              mane_exon_end = as.numeric(unlist(strsplit(EXON_END, split = ',')))), by = .(transcript_id = `#NAME`, gene_name, CHROM = gsub('chr', '', CHROM), STRAND, mane_tx_start = TX_START, mane_tx_end = TX_END)]
spliceai_ann_exons = spliceai_ann[, .(spliceai_exon_start = as.numeric(unlist(strsplit(EXON_START, split = ','))),
                                      spliceai_exon_end = as.numeric(unlist(strsplit(EXON_END, split = ',')))), by = .(gene_name = `#NAME`, CHROM, STRAND, spliceai_tx_start = TX_START, spliceai_tx_end = TX_END)]


# missing genes
comp_ann_genes = unique(spliceai_ann_exons[, .(gene_name, CHROM, STRAND, spliceai_tx_start, spliceai_tx_end)])[unique(mane_ann_exons[, .(gene_name, CHROM, STRAND, mane_tx_start, mane_tx_end)]), on = .(gene_name, CHROM, STRAND)]
#1621 missing genes
missing_genes = comp_ann_genes[is.na(spliceai_tx_start), gene_name]


# missing exons
# annotate with whether they are CDS or UTR (CDS = includes any CDS, even it it's mixed and contains start/stop codon)
# if start or end matches a CDS exon, annotate as CDS
mane_exons = mane_dt[type == 'exon', .(transcript_id, gene_name, seqnames, strand, start, end, exon_number)]
cds_regions = mane_dt[type == 'CDS', .(transcript_id, gene_name, seqnames, strand, start, end, CDS = 1)]

cds_exons = rbind(cds_regions[mane_exons, on = .(transcript_id, gene_name, seqnames, strand, start)][CDS == 1],
                  cds_regions[mane_exons, on = .(transcript_id, gene_name, seqnames, strand, end)][CDS == 1], fill= TRUE)
cds_exons[is.na(i.end), start := i.start]
cds_exons[is.na(i.start), end := i.end]
cds_exons[, `:=` (i.end = NULL, i.start = NULL)]
cds_exons = unique(cds_exons)
mane_exons_info = cds_exons[mane_exons, on = .(transcript_id, gene_name, seqnames, strand, start, end)]
mane_exons_info[is.na(CDS), CDS := 0]

mane_exons_info = mane_exons_info[, .(transcript_id, gene_name, CHROM = gsub('chr', '', seqnames), STRAND = strand, mane_exon_start = start - 1, mane_exon_end= end, CDS, exon_number)]
mane_exons_info[, exon_number := as.numeric(exon_number)]
mane_exons_info[, nexon := max(exon_number), by = .(transcript_id)]
```



```{r}
spliceai_ann_exons[, spliceai_ann_exon := 1]
exon_comp = spliceai_ann_exons[mane_exons_info, on = .(gene_name, CHROM, STRAND, spliceai_exon_start = mane_exon_start, spliceai_exon_end = mane_exon_end)]
# restrict to internal exons
internal_exons_comp = exon_comp[exon_number != 1 & exon_number != nexon]
#internal_exons_comp[!gene_name %in% missing_genes & is.na(spliceai_ann_exon)]
# 1,100 don't have an exact match- how many of these have just one splice-junction changed?

#number of genes containing CDS exons missing
internal_exons_comp[!gene_name %in% missing_genes & is.na(spliceai_ann_exon) & gene_name == 'MYRF' ]
```


# transcripts where boundaries have changed > 50 bp

```{r}
# changed UTR boundaries
# comp_ann_genes[spliceai_tx_start != mane_tx_start & spliceai_tx_end == mane_tx_end]
# comp_ann_genes[spliceai_tx_start == mane_tx_start & spliceai_tx_end != mane_tx_end]
# comp_ann_genes[spliceai_tx_start != mane_tx_start & spliceai_tx_end != mane_tx_end]
#comp_ann_genes[spliceai_tx_start == mane_tx_start & spliceai_tx_end == mane_tx_end]

boundaries_comp = comp_ann_genes[!gene_name %in% missing_genes]

boundaries_comp[, `:=` (tx_start_diff = abs(mane_tx_start - spliceai_tx_start), tx_end_diff = abs(mane_tx_end - spliceai_tx_end))]
# 3,093 where a transcript boundary has changed by > 1000 bases
#boundaries_comp[tx_start_diff > 1000 | tx_end_diff > 1000]
```

```{r}
comparison_summary = data.table(category = c('missing_genes', 'missing_CDS_exons', 'boundary_changes', 'total_genes'),
                                count = c(length(comp_ann_genes[is.na(spliceai_tx_start), gene_name]), 
                                          length(unique(internal_exons_comp[!gene_name %in% missing_genes & is.na(spliceai_ann_exon), gene_name])),
                                          length(unique(boundaries_comp[tx_start_diff > 1000 | tx_end_diff > 1000, gene_name])),
                                          length(unique(c(comp_ann_genes[is.na(spliceai_tx_start), gene_name], 
                                                          internal_exons_comp[!gene_name %in% missing_genes & is.na(spliceai_ann_exon), gene_name],
                                                          boundaries_comp[tx_start_diff > 1000 | tx_end_diff > 1000, gene_name])))))
comparison_summary[, category := factor(category, levels = c('missing_genes', 'missing_CDS_exons', 'boundary_changes', 'total_genes'))]





ggplot(comparison_summary, aes(x = category, y = count, fill = category)) + 
  geom_bar(stat = 'identity', width = 0.5) + 
  xlab('') + scale_fill_brewer(type = 'qual', palette = 8) + 
  theme(legend.position = 'None', 
        axis.text.x = element_text(angle = 20, hjust = 1)) +
  scale_x_discrete(labels = c('Missing genes', 'Genes with missing\nCDS exons', 
                              'Genes with boundary\nchanges > 1000 bp', 'Total genes affected'))
ggsave('figures/transcript_changes.pdf', height = 4, width = 4.5)
```

```{r}
fwrite(comp_ann_genes, 'data/comp_ann_genes.txt', sep = '\t')
fwrite(internal_exons_comp, 'data/internal_exons_comp.txt', sep = '\t')
fwrite(boundaries_comp, 'data/boundaries_comp.txt', sep = '\t')

```



# FIGURE 1A

```{r}
fig_1a = data.table(category = c('precomputed scores', 'updated annotations', 'updated annotations + distance'),
                    variant_count = c(165, 180, 195))

fig_1a[, increased_count := variant_count - 165]

library(tidyverse)
fig_1a[category != 'precomputed scores'] %>%
  ggplot(aes(x = category, y = increased_count))  + 
  geom_bar(stat = 'identity', width = 0.5, fill = 'lightgrey') + 
  ylab('additional variants identified') + xlab('') +
  scale_x_discrete(labels = c('updated transcript annotations', 'updated transcript annotations + \nincreased distance'))
ggsave('figures/increased_variants.pdf', width= 5, height = 6)
```
























