---
title: "R Notebook"
output: 
---

```{r}
library(data.table)
library(tidyverse)
library(ggpubr)
theme_set(theme_classic())
mane_gtf = rtracklayer::import('annotations/MANE.GRCh38.v1.0.ensembl_genomic.gff.gz')
mane_dt = as.data.table(mane_gtf)

mane_ann = fread('annotations/MANEv1.0.txt'  )
spliceai_ann = fread('annotations/grch38.txt')
```


We found that the annotation file used for the precomputed scores included only a short isoform of DNMT3A which excludes this variant. This short isoform excludes 23.3% of the CDS of the MANE isoform, as well as y% of pathogenic ClinVar variants reported in this gene.

```{r}
precomp_dnmt3a_start = spliceai_ann[`#NAME` == 'DNMT3A', TX_START]
precomp_dnmt3a_end = spliceai_ann[`#NAME` == 'DNMT3A', TX_END]

mane_dt[type == 'CDS' & gene_name == 'DNMT3A' & end >  precomp_dnmt3a_end]
sum(mane_dt[type == 'CDS' & gene_name == 'DNMT3A' & end >  precomp_dnmt3a_end, width]) / 
  sum(mane_dt[type == 'CDS' & gene_name == 'DNMT3A', width])


gnomad_dnmt3a = fread('gnomad_data/variants.csv', 
                      col.names = c('clinvar_id', 'variant_id', 
                                    'gold_start', 'clin_significance', 'review_status'))

gnomad_dnmt3a[, var_pos := as.numeric(sapply(strsplit(variant_id, split = '-', fixed = TRUE), '[[', 2))]

nrow(gnomad_dnmt3a[var_pos > precomp_dnmt3a_end  & clin_significance  %in% c("Likely pathogenic", "Pathogenic/Likely pathogenic", "Pathogenic")]) / nrow(gnomad_dnmt3a[ clin_significance  %in% c("Likely pathogenic", "Pathogenic/Likely pathogenic", "Pathogenic")])

```

# overall- get count of missing gens, missing exons (CDS/UTR?), changed CDS exon boundaries, changed UTR boundaries


```{r}
# annotate with gene names
mane_ann = mane_dt[type == 'transcript', .(`#NAME` = transcript_id, gene_name)][mane_ann, on = .(`#NAME`)]


#making table to find missing segemnts in spliceai annotations
mane_ann_exons = mane_ann[, .(mane_exon_start = as.numeric(unlist(strsplit(EXON_START, split = ','))),
                              mane_exon_end = as.numeric(unlist(strsplit(EXON_END, split = ',')))), by = .(transcript_id = `#NAME`, gene_name, CHROM = gsub('chr', '', CHROM), STRAND, mane_tx_start = TX_START, mane_tx_end = TX_END)]
spliceai_ann_exons = spliceai_ann[, .(spliceai_exon_start = as.numeric(unlist(strsplit(EXON_START, split = ','))),
                                      spliceai_exon_end = as.numeric(unlist(strsplit(EXON_END, split = ',')))), by = .(gene_name = `#NAME`, CHROM, STRAND, spliceai_tx_start = TX_START, spliceai_tx_end = TX_END)]


# missing genes
comp_ann_genes = unique(spliceai_ann_exons[, .(gene_name, CHROM, STRAND, spliceai_tx_start, spliceai_tx_end)])[unique(mane_ann_exons[, .(gene_name, CHROM, STRAND, mane_tx_start, mane_tx_end)]), on = .(gene_name, CHROM, STRAND)]
#1621 missing genes
missing_genes = comp_ann_genes[is.na(spliceai_tx_start), gene_name]


# missing exons
# annotate with whether they are CDS or UTR (CDS = includes any CDS, even it it's mixed and contains start/stop codon)
# if start or end matches a CDS exon, annotate as CDS
mane_exons = mane_dt[type == 'exon', .(transcript_id, gene_name, seqnames, strand, start, end, exon_number)]
cds_regions = mane_dt[type == 'CDS', .(transcript_id, gene_name, seqnames, strand, start, end, CDS = 1)]

cds_exons = rbind(cds_regions[mane_exons, on = .(transcript_id, gene_name, seqnames, strand, start)][CDS == 1],
                  cds_regions[mane_exons, on = .(transcript_id, gene_name, seqnames, strand, end)][CDS == 1], fill= TRUE)
cds_exons[is.na(i.end), start := i.start]
cds_exons[is.na(i.start), end := i.end]
cds_exons[, `:=` (i.end = NULL, i.start = NULL)]
cds_exons = unique(cds_exons)
mane_exons_info = cds_exons[mane_exons, on = .(transcript_id, gene_name, seqnames, strand, start, end)]
mane_exons_info[is.na(CDS), CDS := 0]

mane_exons_info = mane_exons_info[, .(transcript_id, gene_name, CHROM = gsub('chr', '', seqnames), STRAND = strand, mane_exon_start = start - 1, mane_exon_end= end, CDS, exon_number)]
mane_exons_info[, exon_number := as.numeric(exon_number)]
mane_exons_info[, nexon := max(exon_number), by = .(transcript_id)]
```



```{r}
spliceai_ann_exons[, spliceai_ann_exon := 1]
exon_comp = spliceai_ann_exons[mane_exons_info, on = .(gene_name, CHROM, STRAND, spliceai_exon_start = mane_exon_start, spliceai_exon_end = mane_exon_end)]
# restrict to internal exons
internal_exons_comp = exon_comp[exon_number != 1 & exon_number != nexon]
#internal_exons_comp[!gene_name %in% missing_genes & is.na(spliceai_ann_exon)]
# 1,100 don't have an exact match- how many of these have just one splice-junction changed?

#number of genes containing CDS exons missing
internal_exons_comp[!gene_name %in% missing_genes & is.na(spliceai_ann_exon) & gene_name == 'MYRF' ]
```


# transcripts where boundaries have changed > 50 bp

```{r}
# changed UTR boundaries
# comp_ann_genes[spliceai_tx_start != mane_tx_start & spliceai_tx_end == mane_tx_end]
# comp_ann_genes[spliceai_tx_start == mane_tx_start & spliceai_tx_end != mane_tx_end]
# comp_ann_genes[spliceai_tx_start != mane_tx_start & spliceai_tx_end != mane_tx_end]
#comp_ann_genes[spliceai_tx_start == mane_tx_start & spliceai_tx_end == mane_tx_end]

boundaries_comp = comp_ann_genes[!gene_name %in% missing_genes]

boundaries_comp[, `:=` (tx_start_diff = abs(mane_tx_start - spliceai_tx_start), tx_end_diff = abs(mane_tx_end - spliceai_tx_end))]
# 3,093 where a transcript boundary has changed by > 1000 bases
#boundaries_comp[tx_start_diff > 1000 | tx_end_diff > 1000]
```

```{r}
comparison_summary = data.table(category = c('missing_genes', 'missing_CDS_exons', 'boundary_changes', 'total_genes'),
                                count = c(length(comp_ann_genes[is.na(spliceai_tx_start), gene_name]), 
                                          length(unique(internal_exons_comp[!gene_name %in% missing_genes & is.na(spliceai_ann_exon), gene_name])),
                                          length(unique(boundaries_comp[tx_start_diff > 1000 | tx_end_diff > 1000, gene_name])),
                                          length(unique(c(comp_ann_genes[is.na(spliceai_tx_start), gene_name], 
                                                          internal_exons_comp[!gene_name %in% missing_genes & is.na(spliceai_ann_exon), gene_name],
                                                          boundaries_comp[tx_start_diff > 1000 | tx_end_diff > 1000, gene_name])))))
comparison_summary[, category := factor(category, levels = c('missing_genes', 'missing_CDS_exons', 'boundary_changes', 'total_genes'))]

```

```{r}
# add wrong ref issues- genes that contain any region of liftover error, so any overlap with the annotation file
wrongref_anno = fread('tool/tool_annotation_files/wrongref_annotation_chr.sorted.txt.gz', col.names = c('CHROM', 'TX_START', 'TX_END', 'recommendation', 'reason'))
setkey(wrongref_anno, CHROM, TX_START, TX_END)
setkey(mane_ann, CHROM, TX_START, TX_END)

wrongref_anno_mane = foverlaps(mane_ann, wrongref_anno, type = 'any', nomatch = NULL)
wrongref_anno_mane[, category := 'liftover error']
```


```{r}
# New figure 1 with 3 bar plots

# plot 1 - number of genes with annotation issues

comparison_summary_data = rbind(comp_ann_genes[is.na(spliceai_tx_start), ] %>% mutate(category = 'missing gene'), 
                                internal_exons_comp[!gene_name %in% missing_genes & is.na(spliceai_ann_exon), ] %>% mutate(category = 'missing CDS exon'),
                                boundaries_comp[tx_start_diff > 1000 | tx_end_diff > 1000, ]%>% mutate(category = 'boundary changes'),
                                wrongref_anno_mane, fill = TRUE)
comparison_summary_data = comparison_summary_data[, .(gene_name, CHROM, STRAND, category, mane_tx_start, mane_tx_end, spliceai_tx_start, spliceai_tx_end, tx_start_diff, tx_end_diff,
                            transcript_id, exon_number, spliceai_exon_start, spliceai_exon_end)]
comparison_summary_plot = comparison_summary_data %>% select(gene_name, category) %>% distinct() %>% group_by(category) %>% tally() %>% setDT()
comparison_summary_plot = rbind(comparison_summary_plot, data.table(category='total', n=length(unique(comparison_summary_data[!is.na(category), gene_name]))))
comparison_summary_plot[, category := factor(category, levels = rev(c('missing gene', 'missing CDS exon', 'boundary changes', 'liftover error', 'total')))]
comparison_summary_plot[, countlabel := scales::comma(n)]

panelb = ggplot(comparison_summary_plot, aes(x = n, y = category, fill = category)) + 
  geom_bar(stat = 'identity', width = 0.5) +  ylab('') +
  xlab('Gene count') + scale_fill_brewer(type = 'qual', palette = 8)+
  geom_text(aes(label = countlabel), 
            position = position_stack(vjust = 1.2))+      
  theme(legend.position = 'None',
    panel.grid.major = element_line(color = "gray", size = 0.15),
    panel.grid.minor = element_line(color = "lightgray", size = 0.1)) +
  scale_y_discrete(labels = rev(c('Missing gene', 'Missing\nCDS exon', 
                              'Gene boundary\nchange > 1000 bp', 'Liftover error', 'Total affected')))+
  scale_x_continuous(labels = scales::comma, breaks = seq(0, 20000, 4000), expand = c(0,0))  + coord_cartesian(xlim = c(0, 10000))
# plot 2 - overlap (proportion?) of panelApp genes with genes with annotation issues
panelapp_genes = fread('panelapp_genes/GREEN_curated_mane_data.gff', col.names = c('chrom', 'a', 'type', 'start', 'end', 'b', 'strand', 'c', 'info', 'gene_name'))
panelapp_genes = panelapp_genes[type == 'gene', ]
comparison_summary_data_panelapp = comparison_summary_data[, .(gene_name, category)][panelapp_genes[, .(gene_name, chrom, start, end, strand)], on = .(gene_name)]
panelapp_props = comparison_summary_data_panelapp %>% select(gene_name, category) %>% distinct() %>% group_by(category) %>% tally() %>% filter(!is.na(category)) %>% setDT()
panelapp_props = rbind(panelapp_props, data.table(category='total', n=length(unique(comparison_summary_data_panelapp[!is.na(category), gene_name]))))
panelapp_props[, total := length(unique(panelapp_genes$gene_name))]
panelapp_props[, prop := n / total]
panelapp_props[, category := factor(category, levels = rev(c('missing gene', 'missing CDS exon', 'boundary changes', 'liftover error', 'total')))]
panelapp_props[, perc := scales::percent(prop)]

panelapp_genes_with_annotation_errors = unique(comparison_summary_data_panelapp[!is.na(category)]$gene_name)

panelapp_genes_with_errors_info = mane_ann[gene_name %in% panelapp_genes_with_annotation_errors] %>% group_by(gene_name, CHROM) %>% dplyr::summarise(start = min(TX_START), end = max(TX_END), .groups = "keep")
fwrite(panelapp_genes_with_errors_info,'data/panelapp_genes_with_annotation_errors.txt', sep = '\t') 

```


```{r}
panelc = ggplot(panelapp_props, aes(x = prop, y = category, fill = category)) + 
  geom_bar(stat = 'identity', width = 0.5) +
  xlab('panelApp green genes (%)') + scale_fill_brewer(type = 'qual', palette = 8) +
  geom_text(aes(label = perc), 
            position = position_stack(vjust = 1.2))+ 
  theme(legend.position = 'None', axis.title.y = element_blank(), axis.text.y = element_blank(),
    panel.grid.major = element_line(color = "gray", size = 0.15),
    panel.grid.minor = element_line(color = "lightgray", size = 0.1)) +
  scale_x_continuous(labels = scales::percent, breaks = seq(0, 1, 0.2), expand = c(0,0)) + coord_cartesian(xlim = c(0, 0.5))

# plot 3 - proportion of theoretical SNVs in precomputed scores overlapping regions with annotation issues
precomp_prop_mock = data.table(category = c('missing gene', 'missing CDS exon', 'boundary changes', 'liftover error'),
                               n = c(188640584, 465516, 95651391, 7128272))
precomp_info = fread('tool/annotation_summary.csv')

precomp_prop = data.table(category = c('missing gene', 'missing CDS exon', 'boundary changes', 'liftover error'),
                          reason = c('missing_gene','missing_exon', 'transcript_boundary_change', 'wrong_REF' ))
precomp_prop = precomp_info[, .(reason, count)][precomp_prop, on = .(reason)]
precomp_prop[, reason := NULL]
precomp_prop = rbind(precomp_prop, data.table(category='total', count=188640584+95651391+465516+1669764))
precomp_prop[, total := precomp_info[reason == 'TOTAL_ORIGINAL', count]]
precomp_prop[, prop := count / total]
precomp_prop[, category := factor(category, levels = rev(c('missing gene', 'missing CDS exon', 'boundary changes', 'liftover error', 'total')))]
precomp_prop[, perc := scales::percent(prop)]

paneld = ggplot(precomp_prop, aes(x = prop, y = category, fill = category)) + 
  geom_bar(stat = 'identity', width = 0.5) +
  geom_text(aes(label = perc), 
            position = position_stack(vjust = 1.2))+ 
  xlab('Theoretical SNVs (%)') + scale_fill_brewer(type = 'qual', palette = 8)  + 
  theme(legend.position = 'None', axis.title.y = element_blank(), axis.text.y = element_blank(),
    panel.grid.major = element_line(color = "gray", size = 0.15),
    panel.grid.minor = element_line(color = "lightgray", size = 0.1)) +
  scale_x_continuous(labels = scales::percent, breaks = seq(0, 1, 0.05), expand = c(0,0)) + coord_cartesian(xlim = c(0, 0.12))

ggarrange(panelb, panelc, paneld, nrow = 1, widths = c(1.5, 1, 1))
ggsave('figures/figure1_raw.pdf', width = 7, height = 3)
```



```{r}
fwrite(comp_ann_genes, 'data/comp_ann_genes.txt', sep = '\t')
fwrite(internal_exons_comp, 'data/internal_exons_comp.txt', sep = '\t')
fwrite(boundaries_comp, 'data/boundaries_comp.txt', sep = '\t')

```


























